[RFC 0097] Unset read permission bit on /nix/store for other users (#97)
NixOS/rfcs: rfcs/0097-no-read-store-dir.md

Fixing the permissions for other files in /proc would likely also be a good idea, regardless of this RFC and regardless
of sandboxing, e.g. it doesn't make sense to expose /proc/config.gz to the entire world even if it doesn't expose any secrets.

Nothing else has to be done likely, since setting the store permissions to `1771` manually doesn't break
anything other than what is mentioned in this document (though it is undone if you run a `nix` command).

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Losing the read (r) bit means that you can't list the files inside the store.
The execute (x) bit allows us to `cd` to it and also access paths inside the store.

E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
inside the store, and not the store itself, but an unprivileged user can't `ls /nix/store` to find the system configuration.

Note: A sandboxed program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
then it could estimate whether the host machine has e.g. Tor on it.

> # Drawbacks

Reviewer: 1. Some users will find it unacceptable that sandboxed programs will be able to probe for specific derivations. For example by probing for common source archives in the store the sandboxed application can likely identify a lot about the host system. This means that this level of access control is not sufficient for some use cases.
2. Leaks are easy to add and may largely remove the benefits of this approach. This makes it a somewhat weak solution as it will be easy to accidentally break the protection and there is no good way to audit against these leaks.

Author: I'm not sure I agree with 2. but 1 is true yeah, for some use cases it isn't enough.
Let's say we're using bwrap to sandbox a program, for example some HTTP server, like so:
```
  env - bwrap
    --unshare-all
    --ro-bind /nix/store /nix/store
    --share-net
    --proc /proc
    --dev /dev
    --dir /tmp
    --setenv TMPDIR /tmp
    --ro-bind /etc/ssl/certs/ca-certificates.crt /etc/ssl/certs/ca-certificates.crt
    --ro-bind /etc/ssl/certs/ca-bundle.crt /etc/ssl/certs/ca-bundle.crt
    --ro-bind /etc/resolv.conf /etc/resolv.conf
    --hostname localhost
    --die-with-parent
    --new-session
    --bind . /data
    --chdir /data
    -- simple-http-server
```
Let's define a leak as the path to a built system configuration, or some similarly "large" derivation such that the closure is essentially everything of importance in the store.
Besides some files in /proc that can be fixed by fixing the permissions, how do you think you could unintentionally expose a leak?

Reviewer: Files in `/etc` can link to fairly large closures, kernel APIs such as `/proc` can also have lots of info, binaries made available to the service may contain this information. The focus of my point is that we have no good way of detecting leaks, this makes me weary of relying on this for fairly untrusted applications.

For example `/proc/*/task/*/mountinfo` is somewhat dangerous and hard to just fix the permissions. Even things like `/proc/sys/kernel/modprobe` could be updated by a user to do something like logging and surprisingly reference a secret.

I'm not saying that this is a blocker. But I think they are important concerns to list so that their cost can be weighed against the benefit of the change.

Author: We're considering this feature in the context of sandboxes, so let's continue from the example I made above (which albeit not completely secure is still good enough for this example).

We don't have access to the entirety of `/etc` in our sandbox, it doesn't make sense in any case. We do not want sandboxed programs to have access to anything in `/etc` that is not relevant to the sandbox. As such, we simply do not mount it.
As for `/proc/*/task/*/mountinfo`, that will inside the sandbox only show what is mounted inside the sandbox.

`/proc/sys` does indeed contain a couple of paths (notably to systemd and modprobe),  but for the sake of argument,
I will assume that there is some global environment that can not be hidden from sandboxes.
The problem is then that the user might not be aware that adding a path to this hypothetical environment will make it accessible to sandboxed programs, which are perhaps handled transparently by NixOS. Then the user could inadvertently add the path to something important, partially ruining the point of sandboxing. However, you could also argue that this behavior would actually be preferred, since what has access to the path should be able to read the path intuitively.

I can not actually think of any such global state that the user would want to add a path to that isn't `/proc/cmdline`, `/proc/config.gz`, etc., but of course that doesn't exclude it from existing.

I will add something about this tomorrow after I wake up (night here in Denmark right now).

A bit off-topic:
Regardless of this RFC, I think it's unfortunate that leaking information into sandboxes is so trivial in Linux.
`/proc/sys/kernel/modprobe` is of no relevance to most programs, yet it is present in `/proc`, which is universally available AFAIK (although namespaced). systemd does have a way of of restricting access to the non-PID related parts of `/proc` through `ProcSubset`, though it says this about it in the man page:
> Since these APIs are used frequently this option is useful only in a few, specific cases, and is not suitable for most non-trivial programs

Reviewer: >  As for /proc/*/task/*/mountinfo, that will inside the sandbox only show what is mounted inside the sandbox.

Yes, but if you map any file from a derivation it will reveal the full path, even if it behind a symlink IIRC. My point isn't that it is impossible to use correctly, but that it is easy to accidentally reveal something that you didn't intend to.

> I will add something about this tomorrow after I wake up (night here in Denmark right now).

Thanks. Again, I don't think this is a blocker as this change does technically improve security. I just want to make sure that we are also documenting the limitations and downside of this approach so that
1. People are extra careful if they do choose to rely upon this change.
2. People don't blindly assume that this is a sufficient solution to remove whatever previous mechanism they have, but instead consider the tradeoff carefully.

Author: I agree, I will add something about it.

Reviewer: I consider this resolved.

[drawbacks]: #drawbacks

> It might be a slight annoyance since shell completion won't work in the /nix/store anymore, e.g.

Reviewer 2: I am constantly relying on shell completion of hashes to investigate what is going on in NixOS, so I view this as a major step back in usability. Maybe there's some other way I could be working, like completing paths base on shell scrollback instead of path completion?

Author: I personally just copy the path in foot (my terminal) by pressing Ctrl+Shift+R.

Reviewer 3: I think I even manage to have workflows that go both ways — sometimes it is nice to complete from a few letters, nicer than doing scrollback search then needing to do selection (or taking the mouse); sometimes it is less nice when zsh rechecks whether intermediate components in the path need expansion on Tab completion

Author: You ought to use foot if possible, it's a great terminal, you basically select what to copy by searching for a prefix of it.

Reviewer 3: That sounds basically what screen does anyway, maybe with more automagic for detecting the end, which will probably sometimes go wrong, even if not too often…

Nah, not even on the same scale as having a chance to affect my willingness of giving up my window manager configuration. (Then again, I don't even use urxvt-specific stuff much and sometimes end up working in fbterm…)

if you have some hash 48914, you can't type `/nix/store/48914<tab>` to get the full path anymore.

External tooling that does a traversal of the nix store (`find`, `du -s`, `ncdu`) would need `sudo` or explicitly given permissions.

# Alternatives
[alternatives]: #alternatives

> If Nix was made to not reset the permissions of /nix/store back to 1775, users who want this change could

Reviewer 3: Or maybe reset to the current mode or-ed with `0o1771`.  That, plus NixOS option to force the mode before running the daemon actually sounds like a first step that would be sufficient for people who actively want this RFC, no? Then people unsure about the implications can try switching completely safely (rolling back a Nix update might hit something unfortunate like a schema change…) And one can run systems differing only in this setting to cross-debug any interesting issues.

Author: I personally don't think it should reset the permissions at all. It is IMO the user's responsibility to set the permissions for /nix/store.

Reviewer 3: Hm, why the RFC proposes to keep the chmod in local-store.cc then?

Author: Because this is simpler, but users should choose themselves, then I think they should have complete freedom wrt. that instead of arbitrarily or'ing with 1771.

Reviewer 3: Intuitively I would expect that the installer ensures the defaults, and then Nix doesn't reset things, which seems to look the same in the normal case but skip one count of «what? why?» if the user actually wants to change the permissions. 

So in the first step there is not even a need to change the installer at all (but maybe release notes should include a mention)

---
feature: no-read-store-dir
start-date: 2021-07-04
author: Las Safin
co-authors:

> shepherd-team:

Reviewer 4: ```suggestion
shepherd-team: @kevincox @7c6f434c @edolstra 
shepherd-leader: @kevincox
```

> shepherd-leader: @kevincox

Reviewer 5: ```suggestion
shepherd-leader: @edolstra
```

https://github.com/NixOS/rfcs/pull/97#issuecomment-916076773

Author: Didn't notice that the elected shepherd leader had changed... https://github.com/NixOS/rfcs/pull/97#discussion_r701091017

Reviewer 4: @L-as I think we simply missed your yaml assignment when discussing this RFC.

related-issues:
---

# Summary
[summary]: #summary

Set the permissions for `/nix/store`, `/nix/store/.links` and `/nix/store/trash` to 1771 instead of 1775, disabling reading the directory for other users.

# Motivation
[motivation]: #motivation

This is a simple change that acts as an extra layer of security by making
it harder to access store paths that programs don't need to access.

For a directory, being able to execute without being able to read it means
that you can access paths inside the directory, if and only if you know the path beforehand,
since you can not list the directory entries.

In our case, with this change, all users that are not root and are not part of nixbld,
will only be able to directly access paths inside `/nix/store` which they already know,
i.e. the hash and the name.
If you run some program that does not have access to the nix daemon under such

> a user, it will not be able to search the store for built system configurations,

Reviewer 6: What exactly do you mean with "search the store" and do applications actually do this? Because I read this as "applications that scrape the store for configuration" and I'm pretty sure I got that wrong.

Author: I know of no applications that do this; it is just an example. So yes, applications that scrape the store for configuration, which is what doing `ls /nix/store` is. This is what this RFC is about, scraping the store.

Reviewer 6: Then I suggest removing that line and the following one. There are many legitimate use cases for listing the store (and they'll break and they should be mentioned + alternatives), but "looking for configuration" totally isn't one.

Author: That is already in the RFC under Drawbacks.

will be, and they will likely not have the same considerations for "paths as secrets" as us.
Essentially, you have to consider the store paths as encodings of their contents wrt. sandboxes.

The above points about sandboxed programs are not negative effects of changing the permissions of /nix/store and such,
but rather considerations for making use of this change in the context of sandboxes.

# Alternatives
[alternatives]: #alternatives

If Nix was made to not reset the permissions of /nix/store back to 1775, users who want this change could
do it themselves by simply putting this into their configuration.nix:
```nix
{
  system.activationScripts.chmod-store.text = ''
    ${pkgs.util-linux}/bin/unshare -m ${pkgs.bash}/bin/sh -c '${pkgs.util-linux}/bin/mount -o remount,rw /nix/store ; ${pkgs.coreutils}/bin/chmod 1771 /nix/store ; ...'
  '';
}
```

Traditional sandboxing could be seen as an alternative to this, however, they are not mutually exclusive,
since this change is fundamentally just changing some permissions in NixOS.
Improving the permissions for /nix/store will not decrease security for traditional sandboxes.

> # Unresolved questions

Reviewer 6: How does this affect other use cases of Nix, like on non-NixOS or single-user installations?

Author: It won't be any different. Of course, on non-NixOS systems the permissions of /proc/cmdline, etc. are the responsibility of that distro, but everything else should be exactly the same.

---
feature: nix-store-perms
start-date: 2021-07-04
author: Las Safin
co-authors:
shepherd-team: @kevincox @7c6f434c @edolstra
shepherd-leader: @edolstra
related-issues:
---

# Summary
[summary]: #summary

- NixOS should have a module for configuring the permissions set for `/nix/store` on boot.
- Nix should not enforce the permissions used for `/nix/store`.
- The default permissions if the store doesn't exist should be 1735 when the store is made by Nix or the NixOS installer.
  This means that the nixbld group can't `ls` the directory.

# Motivation
[motivation]: #motivation

Right now you can't set the permissions for `/nix/store`, since they'll be overwritten
by Nix anytime you use `nix`.

> We want to `chmod g-r /nix/store`, because the `nixbld` group doesn't actually

Reviewer: ```suggestion
`chmod g-r /nix/store` is beneficial because the `nixbld` group doesn't actually
```

Nit: Avoid first-person pronouns.

need to read the directory. It only needs to be able to write and "execute" it.

> This, however, should be optional, since the user should be able to do what they want.

Reviewer: ```suggestion
This, however, should be optional, since the user should be able to configure the permissions however they want.
```

Nit: define the subject more clearly.

Some users might also want to do things like `chmod o-r /nix/store`, which
gives you the interesting property that you can not access paths you do not
already know of.
Do note that given that all processes can by default read `/proc/cmdline`,

> `/run/current-system`, and many other places, they can still read your
> system's closure, which makes it an insufficient solution for security in many cases.

Reviewer: ```suggestion
`/run/current-system` and many other places which reveals your system's closure, making this permission change an insufficient solution for security in many cases.
```

Specifically, we need to modify the following places (not exhaustive):
- [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62)
- [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577)
- [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278)
- [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641)
- [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

You should be able to do something like the following:
```nix
nix.store-perms = "xxxx";
```

# Drawbacks
[drawbacks]: #drawbacks

If a user on a non-NixOS platform mistakenly sets the permissions for `/nix/store` to

> something else, it won't be reverted by Nix automatically.

Reviewer: ```suggestion
something undesirable, it won't be reverted by Nix automatically.
```

Nit: Wording.

The fact that you can set the permissions to something "else" is the whole point.

- Nix should not enforce the permissions used for `/nix/store`.
- The default permissions if the store doesn't exist should be 1735 when the store is made by Nix or the NixOS installer.
  This means that the nixbld group can't `ls` the directory.

# Motivation
[motivation]: #motivation

Right now you can't set the permissions for `/nix/store`, since they'll be overwritten
by Nix anytime you use `nix`.

`chmod g-r /nix/store` is beneficial because the `nixbld` group doesn't actually
need to read the directory. It only needs to be able to write and "execute" it.
This, however, should be optional, since the user should be able to configure
the permissions however they want.

Some users might also want to do things like `chmod o-r /nix/store`, which
gives you the interesting property that you can not access paths you do not
already know of.
Do note that given that all processes can by default read `/proc/cmdline`,
`/run/current-system`, and many other places which reveal your
system's closure, making this permission change an insufficient solution for
security in many cases. This, however, is also entirely optional and is not

> the default in any way.

Reviewer 7: I assume that at least some filesystems are vulnerable to timing attacks, probing paths with `stat` to build prefixes of subdirectory paths until the full path is recovered; using the `x` bit to recover `r` for all paths, including ones that aren't running.

Nonetheless, it's good for practical hermeticity.

Reviewer 3: Yes, we hope that people either look at the list and start, like you, name extra leak vectors, or trust «insufficient for security» claim. It would be interesting to have a survey of path leak vectors, but this RFC is shorter than any useful survey like that…

Reviewer: I agree. Any list we try to create in this RFC is likely incomplete so it is better to just state that as an assumption and accept it for the rationale here. If someone wants to prove that the list of leaks is empty (and expected to stay that way) they can propose a follow-up RFC to rely on that fact.

Specifically, we need to modify the following places (not exhaustive):
- [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62)
- [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577)
- [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278)
- [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641)
- [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

You should be able to do something like the following:
```nix
nix.store-perms = "xxxx";
```

# Drawbacks
[drawbacks]: #drawbacks

If a user on a non-NixOS platform mistakenly sets the permissions for `/nix/store` to

> something undesirable, it won't be reverted by Nix automatically.

Reviewer 7: Why not restrict it to sensible or at least not terrible values? Seems quite feasible and the allow list could be extended if we discover a new use case.

Reviewer 3: Because this «undesirable» includes the idea that most likely unfortunate mistakes are someon else's «just as planned» (I am not even sure 0000 is always useless; hmmm, now I am tempted to try once this is merged and trying is easy)

# Alternatives
[alternatives]: #alternatives

You could not do this and keep it as it is.

# Unresolved questions
[unresolved]: #unresolved-questions

There doesn't seem to be any.

# Future work
[future]: #future-work

> In the future we likely want to reduce the default permissions for `/nix/store` as much as possible.

Reviewer 7: Sometimes I use tab completion on a store path; saves some mousing when I fail to copy a path completely, or sometimes typing three or four letters of the hash is easy enough.

---
feature: no-read-store-dir
start-date: 2021-07-04
author: Las Safin
co-authors:
shepherd-team:
shepherd-leader:
related-issues:
---

# Summary
[summary]: #summary

Set the permissions for /nix/store to 1771 instead of 1775, disabling reading the directory for other users.

# Motivation
[motivation]: #motivation

This means that you can not trivially see all the paths in the nix store, e.g. `ls` won't work
on /nix/store without sudo unless you're in the nixbld group.

Almost everything in NixOS needs access to /nix/store, which means that all your systemd services,
your flatpak programs, your manually [bubblewrap](https://github.com/containers/bubblewrap)ed programs
will almost certainly have access to /nix/store and see all of its contents.
That means they will see your NixOS configuration, initrd secrets (unless your bootloader has support for it),

> all the programs you've built, your usage activity, and possibly also your important secrets if you

Reviewer 8: What does `usage activity` mean here?

If it's about discovering which programs a user has installed, wouldn't that still be possible for a malicious containerised app that cannot list `/nix/store` but check the presence of known store paths, by constructing externally the last 10 years of output store paths built by Hydra, and `stat()`ing them on the local machine to check if they exist?

For example, such a program could check whether you have `tor-browser` installed by collecting the ~1000 `*-tor-browser` store paths that Hydra built, and check if you have any of them in the store.

Author: You are right, I did not consider that.

Set the 1775
in [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62),
in [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577), and
in [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278)
to 1771.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Losing the read (r) bit means that you can't list the files inside the store.
The execute (x) bit allows us to `cd` to it and also access paths inside the store.

E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory
inside the store, and not the store itself, but you can't without sudo do `ls /nix/store` to find your system configuration.

# Drawbacks
[drawbacks]: #drawbacks

It might be a slight annoyance since shell completion won't work in the /nix/store anymore, e.g.
if you have some hash 48914, you can't type `/nix/store/48914<tab>` to get the full path anymore.

> 

Reviewer 8: ```suggestion

External tooling that does a traversal of the nix store (`find`, `du -s`, `ncdu`) will need `sudo` or explicitly given permissions.

```

on /nix/store without sudo unless you're in the nixbld group.

Almost everything in NixOS needs access to /nix/store, which means that all your systemd services,
your flatpak programs, your manually [bubblewrap](https://github.com/containers/bubblewrap)ed programs
will almost certainly have access to /nix/store and see all of its contents.
That means they will see your NixOS configuration, initrd secrets (unless your bootloader has support for it),
all the programs you've built, and possibly also your important secrets if you
(unfortunately) had to put them in the store!

By simply removing the ability to read /nix/store (but not execute!), programs will only be able
to access the store paths *which they already know the hash and name of*. This is a huge boon to security
even if you don't have secrets in your store, as it will almost completely eliminate the above
problems. That isn't to say it's a complete solution to security, but it will *allow* much more
complete solutions to security, for example with bubblewrap, meaning that you can be sure that
your sandboxed game won't suddenly scrape your store and send it off to NSA.

> # Detailed design

Reviewer: Have you tested this design and does it appear to work with no further changes? That seems to be implied but it would be nice to note that explicitly.

Author: Yes it does work, until you run some `nix` command which will set it back to 1775.

[design]: #detailed-design

Set the 1775
in [nixpkgs/nixos/modules/system/boot/stage-2-init.sh](https://github.com/NixOS/nixpkgs/blob/8284fc30c84ea47e63209d1a892aca1dfcd6bdf3/nixos/modules/system/boot/stage-2-init.sh#L62),
in [nix/scripts/install-multi-user.sh](https://github.com/NixOS/nix/blob/cf1d4299a8fa8906f62271dcd878018cef84cc30/scripts/install-multi-user.sh#L577),
in [nix/src/libstore/globals.hh](https://github.com/NixOS/nix/blob/ba8b39c13003c8ddafb6bec308997e09b9851c46/src/libstore/globals.hh#L278),
in [nix/src/libstore/build/local-derivation-goal.cc](https://github.com/NixOS/nix/blob/6182ae689826554d915b4ed72e07f7978dc1d13c/src/libstore/build/local-derivation-goal.cc#L641), and
in [nix/src/libstore/local-store.cc](https://github.com/NixOS/nix/blob/0a535dd5ac93576f7152d786464e330ae3d46b50/src/libstore/local-store.cc#L181)
to 1771.

# Examples and Interactions
[examples-and-interactions]: #examples-and-interactions

Losing the read (r) bit means that you can't list the files inside the store.
The execute (x) bit allows us to `cd` to it and also access paths inside the store.

E.g. `ls "$(readlink /nix/var/nix/profiles/system)"` will still work, since this is a directory

> inside the store, and not the store itself, but you can't without sudo do `ls /nix/store` to find your system configuration.

Reviewer: ```suggestion
inside the store, and not the store itself, but an unprivileged user can't `ls /nix/store` to find the system configuration.
```

Note: A program could still have some idea of how the host machine is used by checking each store path that Hydra has ever built,
then it could estimate whether the host machine has e.g. Tor on it.

# Drawbacks
[drawbacks]: #drawbacks

It might be a slight annoyance since shell completion won't work in the /nix/store anymore, e.g.
if you have some hash 48914, you can't type `/nix/store/48914<tab>` to get the full path anymore.

External tooling that does a traversal of the nix store (`find`, `du -s`, `ncdu`) would need `sudo` or explicitly given permissions.

# Alternatives
[alternatives]: #alternatives

I had [this script](https://github.com/L-as/nix-misc/blob/e844a03ebf4cad4fc8eca0e52306788b70c2a60d/claybox.rb) that is a wrapper around bubblewrap before, but doing this
is a lot cleaner, is system-wide, and is a lot faster, since bind mounting each individual store path with bubblewrap can be quite slow
if you have many (it's O(n^2) IIRC).

Another alternative for users could be just doing `sudo unshare -m bash -c 'mount -o remount,rw /nix/store; chmod 1771 /nix/store'`,
but this doesn't work right now because all of the `nix` executables set it to 1775 again.

> # Unresolved questions

Reviewer: What is the stance with "leaks" like `/nix/var/nix/profiles/system`? Should we expect to remove them over time or are they expected to live forever? While this will give some security (for example I build a derivation in a private directory others won't be able to find it) if we have a number of leaks that provide access to a huge swath of the derivations what is the point.

Other leaks that I am aware of:

- `/proc/cmdline`
- `/proc/{pid}/*`

Author: You'd disallow access to `/nix/var` instead of removing `/nix/var` (which would make no sense). This can be done through making a separate user namespace, then a separate mount namespace AFAIK. `/proc/{pid}/*` wouldn't be a separate problem if you make a separate pid namespace. This would basically just entail using bubblewrap or some other wrapper tool.

`/proc/cmdline` is indeed problematic, since it's available in all pid namespaces for some reason, you could however pass the system configuration to boot by making it a part of the initrd rather than choosing it by a kernel command-line parameter.

As I note in the document, the actual improvement only happens if you remove the leaks by using it in combination with some other sandboxing mechanism.
Essentially, with this system, knowing the hash of some data also means knowing the data itself. Thus, it isn't a huge improvement by itself, but it is a small incremental hardening with few drawbacks that allows for better security.

Reviewer: Ah, so the immediate goal isn't to make the system secure against ordinary users, but make it easier to create effective sandboxes. Is thay understanding correct?

Author: Yes, that is correct.

Author: Do you have any recommendations for how to make this more clear in the document?

Author: How is the new explanation?

Author: I actually found out that /proc/cmdline has the same permissions across all pid namespaces apparently, so just fixing the permissions should be enough!

Reviewer: Thanks, I think it is clear now that the immediate benefits are primarily for sandboxed processes.

---
feature: no-read-store-dir
start-date: 2021-07-04
author: Las Safin
co-authors:
shepherd-team:
shepherd-leader:
related-issues:
---

# Summary
[summary]: #summary

> Set the permissions for /nix/store to 1771 instead of 1775, disabling reading the directory for other users.

Reviewer 9: Don't forget about `/nix/store/.links` (The storage location for optimized files) and `/nix/store/trash` (Used to stage directories for GC).

Author: Thanks.
